subprojects/core/src/main/java/org/gradle/process/internal/daemon/AbstractWorkerDaemonExecutor.java

```diff
@@ -0,0 +1,145 @@
+    }
+
+    @Override
+    public WorkerDaemonExecutor<T> forkOptions(Action<JavaForkOptions> forkOptionsAction) {
+        forkOptionsAction.execute(javaForkOptions);
+        return this;
+    }
+
+    @Override
+    public JavaForkOptions getForkOptions() {
+        return javaForkOptions;
+    }
+
+    @Override
+    public WorkerDaemonExecutor<T> params(Serializable... params) {
+        this.params = params;
+        return this;
+    }
+
+    protected Set<File> getClasspath() {
+        return classpath;
+    }
+
+    protected Set<String> getSharedPackages() {
+        return sharedPackages;
+    }
+
+    protected Class<? extends T> getImplementationClass() {
+        return implementationClass;
+    }
+
+    protected Serializable[] getParams() {
+        return params;
+    }
+
+    protected WorkerDaemonFactory getWorkerDaemonFactory() {
+        return workerDaemonFactory;
+    }
+
+    protected Class<? extends WorkerDaemonProtocol> getServerImplementationClass() {
+        return serverImplementationClass;
+    }
+
+    protected DaemonForkOptions getDaemonForkOptions() {
+        Iterable<Class<?>> paramTypes = CollectionUtils.collect(getParams(), new Transformer<Class<?>, Object>() {
+            @Override
+            public Class<?> transform(Object o) {
+                return o.getClass();
+            }
+        });
+        return toDaemonOptions(implementationClass, paramTypes, javaForkOptions, classpath, sharedPackages);
+    }
+
+    private DaemonForkOptions toDaemonOptions(Class<?> actionClass, Iterable<Class<?>> paramClasses, JavaForkOptions forkOptions, Iterable<File> classpath, Iterable<String> sharedPackages) {
+        ImmutableSet.Builder<File> classpathBuilder = ImmutableSet.builder();
+        ImmutableSet.Builder<String> sharedPackagesBuilder = ImmutableSet.builder();
+
+        if (classpath != null) {
+            classpathBuilder.addAll(classpath);
+        }
+
+        if (sharedPackages != null) {
+            sharedPackagesBuilder.addAll(sharedPackages);
+        }
+
+        addVisibilityFor(actionClass, classpathBuilder, sharedPackagesBuilder);
+
+        for (Class<?> paramClass : paramClasses) {
+            addVisibilityFor(paramClass, classpathBuilder, sharedPackagesBuilder);
+        }
+
+        Iterable<File> daemonClasspath = classpathBuilder.build();
+        Iterable<String> daemonSharedPackages = sharedPackagesBuilder.build();
+
+        return new DaemonForkOptions(forkOptions.getMinHeapSize(), forkOptions.getMaxHeapSize(), forkOptions.getAllJvmArgs(), daemonClasspath, daemonSharedPackages);
+    }
+
+    private static void addVisibilityFor(Class<?> visibleClass, ImmutableSet.Builder<File> classpathBuilder, ImmutableSet.Builder<String> sharedPackagesBuilder) {
+        if (visibleClass.getClassLoader() != null) {
+            classpathBuilder.addAll(ClasspathUtil.getClasspathFiles(visibleClass.getClassLoader()));
+        }
+
+        if (visibleClass.getPackage() != null) {
```

Comment:
I may not be thinking about this correctly, but I think we at least need to set up the shared package for the Runnable to be executed.  Something has to reach from the implementation classpath to the application classpath to instantiate the Runnable and needs to be able to get past the `FilteringClassLoader` in between.  Having said that, I think I can clean this class up a little and _only_ add the package for the Runnable, which is maybe what you meant here.

Or are you suggesting that we put _everything_ in the implementation classloader?